lib/matplotlib/backends/backend_pgf.py

```diff
@@ -254,21 +254,21 @@ def _cleanup_remaining_instances():
             latex_manager._cleanup()
 
     def _stdin_writeln(self, s):
-        self.latex_stdin_utf8.write(s)
-        self.latex_stdin_utf8.write("\n")
-        self.latex_stdin_utf8.flush()
+        self.latex.stdin.write(s)
+        self.latex.stdin.write("\n")
+        self.latex.stdin.flush()
 
     def _expect(self, s):
-        exp = s.encode("utf8")
-        buf = bytearray()
+        s = list(s)
+        buf = []
```

Comment:
Why do we need a list of chars that are joined later? Can‘t we make s a str and append to tha directly?

Reply:
IIRC, appending to a string suffers from a line painters problem where you get n^2 scaling due to having to copy the string when you append to it.  When you use `''.join` python does some clever stuff under the hood to pre-allocate the right size memory buffer and and only copy your fragments once.

```
In [3]: def appender(): 
   ...:     s = '' 
   ...:     for j in range(1000): 
   ...:         s += 'a' 
   ...:                                                                                                                                                                                                                               

In [4]: def collector(): 
   ...:     ''.join(['a' for j in range(1000)]) 
   ...:                                                                                                                                                                                                                               

In [5]: %timeit appender()                                                                                                                                                                                                            
88.7 µs ± 636 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [6]: %timeit collector()                                                                                                                                                                                                           
45.2 µs ± 897 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [7]:     

```

Reply:
Indeed, that's the reason.

Reply:
I see. In that case, I'd call `buf`, `chars` instead, because that's what it is, a collection of characters.

Reply:
sure, done